saleor/graphql/discount/mutations/sale/sale_channel_listing_update.py

```diff
@@ -73,18 +82,75 @@ class Meta:
             channel = add_channel["channel"]
-            defaults = {"currency": channel.currency_code}
-            channel = add_channel["channel"]
-            if "discount_value" in add_channel.keys():
-                defaults["discount_value"] = add_channel.get("discount_value")
-            SaleChannelListing.objects.update_or_create(
-                sale=sale,
-                channel=channel,
-                defaults=defaults,
+            discount_value = add_channel["discount_value"]
+
+            if channel.id not in current_channel_ids:
+                rules_data_to_add.append(
+                    RuleInfo(
+                        rule=PromotionRule(
+                            name=examplart_rule.name,
+                            promotion=promotion,
+                            catalogue_predicate=examplart_rule.catalogue_predicate,
+                            reward_value_type=examplart_rule.reward_value_type,
+                            reward_value=discount_value,
+                        ),
+                        channel=channel,
+                    )
+                )
+            else:
+                # We ensure that every rule has one or none related channel
+                rule_to_update = rules.get(
+                    id=rule_channel.get(channel_id=channel.id).promotionrule_id  # type: ignore[attr-defined] # noqa: E501
+                )
+                rule_to_update.reward_value = discount_value
+                rules_to_update.append(rule_to_update)
+
+        for rule_data in rules_data_to_add:
+            rule_data.rule.assign_old_channel_listing_id()
+
+        new_rules = [rule_data.rule for rule_data in rules_data_to_add]
+        PromotionRule.objects.bulk_create(new_rules)
+        rules_channels = [
+            PromotionRuleChannel(
+                promotionrule=rule_data.rule, channel=rule_data.channel
             )
+            for rule_data in rules_data_to_add
+        ]
+        PromotionRuleChannel.objects.bulk_create(rules_channels)
+        PromotionRule.objects.bulk_update(rules_to_update, ["reward_value"])
+
+    @classmethod
+    def remove_channels(cls, promotion: Promotion, remove_channels: List[int]):
+        rules = promotion.rules.all()
+        PromotionRuleChannel = PromotionRule.channels.through
+        rule_channel = PromotionRuleChannel.objects.filter(
+            channel_id__in=remove_channels
+        ).filter(Exists(rules.filter(id=OuterRef("promotionrule_id"))))
+        if not rule_channel:
+            return
+        rules_to_delete_ids = list(
+            rule_channel.values_list("promotionrule_id", flat=True)
+        )
+        # We ensure at least one rule is assigned to promotion in order to
+        # determine old sale's type and catalogue
+        if len(rule_channel) >= len(rules):
+            rule_left_id = rules_to_delete_ids.pop()
+            rule_channel.filter(promotionrule_id=rule_left_id).delete()
```

Comment:
But we should remove channels from all rules

Reply:
We do it in the next line, by deleting all the rules, except `rule_left_id`. This way we get rid of all channel listing, but keep a single rule (without channel).